Add MKLDNN backend to `matmul` for `bfloat32`

zhuhaozhe opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116015

### Testing
FP32 matmul vs. mkldnn BF32 matmul  on SPR

single core:

Input | BF32   / ms | FP32  /   ms | Speed up
-- | -- | -- | --
M: 128, N: 128, K: 128, trans_a: False, trans_b: False | 32.842 | 38.279 | 1.165
M: 128, N: 256, K: 128, trans_a: False, trans_b: False | 38.590 | 73.967 | 1.917
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 18456.267 | 74588.002 | 4.041


56 cores:
Input | BF32   / ms | FP32 /   ms | Speed up
-- | -- | -- | --
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 1199.400 | 1715.548 | 1.430
M: 8192, N: 768, K: 768, trans_a: False, trans_b: True |1129.204 | 1708.912 |  1.513
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: False | 3655.915  | 7992.877 | 2.186
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: True | 3707.993 |  8026.191 | 2.165
Batch: 768, M: 128, N: 64, K: 128  | 1296.419 | 1308.411 | 1.009

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

zhuhaozhe committed:
enable mkldnn bf32

[ghstack-poisoned]

zhuhaozhe committed:
Update on "[WIP] enable mkldnn bf32"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on torch/testing/_internal/common_mkldnn.py:
```diff
@@ -0,0 +1,85 @@
+
+
+@contextlib.contextmanager
+def bf32_on(self, bf32_precision=1e-5):
```
May I know how `bf32_precision` is applied to the test?

zhuhaozhe replied:
Yes, with this https://github.com/pytorch/pytorch/pull/116015/files/f6f4686015a4a8795fd5305473cc232db79abc4c#diff-7e17421f32124016eb8de04dc2f445da5786a28355e1addc72b305466f590180R609 will run bf32_on (following `with_tf32_on_and_off`)

jgong5 replied:
This doesn't seem correct. `tf32_on` changes `self.precision` per `tf32_precision` arg (https://github.com/pytorch/pytorch/blob/e3aefe297050e197ff3f36f555a134a0e61ffc9c/torch/testing/_internal/common_cuda.py#L94). But you don't. Why?

zhuhaozhe replied:
Thanks for the reminder! @jgong5 , I misunderstood your question before. Now I have fixed it.

jgong5 commented on aten/src/ATen/Context.cpp:
```diff
@@ -44,6 +44,14 @@ void Context::setUserEnabledMkldnn(bool e) {
   enabled_mkldnn = e;
 }
 
+bool Context::allowBF32MkldnnMatmul() const {
+  return float32_matmul_precision == at::Float32MatmulPrecision::MEDIUM;
+}
+
+void Context::setAllowBF32MkldnnMatmul(bool b) {
+  float32_matmul_precision = b ? at::Float32MatmulPrecision::MEDIUM : float32_matmul_precision;
+}
```
Can we reuse `setFloat32MatmulPrecision` and `float32MatmulPrecision` without introducing new backend functions? Also suggest not to introduce the new "bf32" concept to the end users.

zhuhaozhe replied:
Hi, @jgong5. This is following `CuBLAS`.
`CuBLAS TF32` can be enabled by either `setFloat32MatmulPrecision('high')` or `torch._C._set_cublas_allow_tf32(value)`.
So I make `MKLDNN BF32` can be enabled can be enabled by either `setFloat32MatmulPrecision('medium')` or `torch._C._set_mkldnn_bf32_matmul_enabled(True)`.

If `BF32` is not suggest to introduced to end users like `TF32`, I can remove `torch._C._set_mkldnn_bf32_matmul_enabled(True)`. and only use `setFloat32MatmulPrecision('medium')`.
Do you have more preference on it?

jgong5 replied:
> and only use `setFloat32MatmulPrecision('medium')`.
> Do you have more preference for it?

Yes, using `setFloat32MatmulPrecision('medium')` sounds better.

zhuhaozhe replied:
Hi, @jgong5. I have removed `mkldnn_bf32_matmul_enabled` `get/set`

zhuhaozhe committed:
Update aten/src/ATen/native/mkldnn/Matmul.cpp

Co-authored-by: Jiong Gong <jiong.gong@intel.com>

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "enable mkldnn bf32 matmul"


### Testing
FP32 matmul vs. mkldnn BF32 matmul  on SPR

single core:

Input | BF32   / ms | FP32  /   ms | Speed up
-- | -- | -- | --
M: 128, N: 128, K: 128, trans_a: False, trans_b: False | 32.842 | 38.279 | 1.165
M: 128, N: 256, K: 128, trans_a: False, trans_b: False | 38.590 | 73.967 | 1.917
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 18456.267 | 74588.002 | 4.041


56 cores:
Input | BF32   / ms | FP32 /   ms | Speed up
-- | -- | -- | --
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 1199.400 | 1715.548 | 1.430
M: 8192, N: 768, K: 768, trans_a: False, trans_b: True |1129.204 | 1708.912 |  1.513
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: False | 3655.915  | 7992.877 | 2.186
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: True | 3707.993 |  8026.191 | 2.165
Batch: 768, M: 128, N: 64, K: 128  | 1296.419 | 1308.411 | 1.009

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable mkldnn bf32 matmul"


### Testing
FP32 matmul vs. mkldnn BF32 matmul  on SPR

single core:

Input | BF32   / ms | FP32  /   ms | Speed up
-- | -- | -- | --
M: 128, N: 128, K: 128, trans_a: False, trans_b: False | 32.842 | 38.279 | 1.165
M: 128, N: 256, K: 128, trans_a: False, trans_b: False | 38.590 | 73.967 | 1.917
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 18456.267 | 74588.002 | 4.041


56 cores:
Input | BF32   / ms | FP32 /   ms | Speed up
-- | -- | -- | --
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 1199.400 | 1715.548 | 1.430
M: 8192, N: 768, K: 768, trans_a: False, trans_b: True |1129.204 | 1708.912 |  1.513
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: False | 3655.915  | 7992.877 | 2.186
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: True | 3707.993 |  8026.191 | 2.165
Batch: 768, M: 128, N: 64, K: 128  | 1296.419 | 1308.411 | 1.009

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update __init__.py

jgong5 commented on aten/src/ATen/native/mkldnn/Matmul.cpp:
```diff
@@ -53,8 +53,25 @@ bool mkldnn_fp16_gemm(
     c10::Half *c, int64_t ldc) {
   return false;
 }
+bool mkldnn_bf32_gemm(
```
Suggest to add a note somewhere in the code to explain what "bf32" mean.

zhuhaozhe replied:
Added.

zhuhaozhe committed:
Update on "enable mkldnn bf32 matmul"


### Testing
FP32 matmul vs. mkldnn BF32 matmul  on SPR

single core:

Input | BF32   / ms | FP32  /   ms | Speed up
-- | -- | -- | --
M: 128, N: 128, K: 128, trans_a: False, trans_b: False | 32.842 | 38.279 | 1.165
M: 128, N: 256, K: 128, trans_a: False, trans_b: False | 38.590 | 73.967 | 1.917
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 18456.267 | 74588.002 | 4.041


56 cores:
Input | BF32   / ms | FP32 /   ms | Speed up
-- | -- | -- | --
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 1199.400 | 1715.548 | 1.430
M: 8192, N: 768, K: 768, trans_a: False, trans_b: True |1129.204 | 1708.912 |  1.513
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: False | 3655.915  | 7992.877 | 2.186
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: True | 3707.993 |  8026.191 | 2.165
Batch: 768, M: 128, N: 64, K: 128  | 1296.419 | 1308.411 | 1.009

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/mkldnn/Matmul.h:
```diff
@@ -48,6 +48,11 @@ bool mkldnn_fp16_gemm(
     c10::Half *c, int64_t ldc);
 
+/*
+oneDNN implicit reduced precision arithmetic feature
+https://github.com/mgouicem/oneDNN/tree/mgouicem/rfcs/implicit_downconvert/rfcs/20210301-computation-datatype
+to allow implicitly cast data type from FP32 to BF16 in onednn compute primitives
```
```suggestion
oneDNN implicit reduced precision arithmetic feature
https://github.com/mgouicem/oneDNN/tree/mgouicem/rfcs/implicit_downconvert/rfcs/20210301-computation-datatype
to allow implicitly cast data type from FP32 to BF16 in onednn compute primitives. We call it
"BF32" mode here to call out its difference from FP32 and BF16 compute.
```

zhuhaozhe committed:
Update __init__.py

zhuhaozhe committed:
Update aten/src/ATen/native/mkldnn/Matmul.h

Co-authored-by: Jiong Gong <jiong.gong@intel.com>

zhuhaozhe commented:
@pytorchbot merge

zhuhaozhe commented:
@pytorchbot rebase

zhuhaozhe committed:
Update on "enable mkldnn bf32 matmul"


### Testing
FP32 matmul vs. mkldnn BF32 matmul  on SPR

single core:

Input | BF32   / ms | FP32  /   ms | Speed up
-- | -- | -- | --
M: 128, N: 128, K: 128, trans_a: False, trans_b: False | 32.842 | 38.279 | 1.165
M: 128, N: 256, K: 128, trans_a: False, trans_b: False | 38.590 | 73.967 | 1.917
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 18456.267 | 74588.002 | 4.041


56 cores:
Input | BF32   / ms | FP32 /   ms | Speed up
-- | -- | -- | --
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 1199.400 | 1715.548 | 1.430
M: 8192, N: 768, K: 768, trans_a: False, trans_b: True |1129.204 | 1708.912 |  1.513
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: False | 3655.915  | 7992.877 | 2.186
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: True | 3707.993 |  8026.191 | 2.165
Batch: 768, M: 128, N: 64, K: 128  | 1296.419 | 1308.411 | 1.009

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable mkldnn bf32 matmul"


### Testing
FP32 matmul vs. mkldnn BF32 matmul  on SPR

single core:

Input | BF32   / ms | FP32  /   ms | Speed up
-- | -- | -- | --
M: 128, N: 128, K: 128, trans_a: False, trans_b: False | 32.842 | 38.279 | 1.165
M: 128, N: 256, K: 128, trans_a: False, trans_b: False | 38.590 | 73.967 | 1.917
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 18456.267 | 74588.002 | 4.041


56 cores:
Input | BF32   / ms | FP32 /   ms | Speed up
-- | -- | -- | --
M: 8192, N: 768, K: 768, trans_a: False, trans_b: False | 1199.400 | 1715.548 | 1.430
M: 8192, N: 768, K: 768, trans_a: False, trans_b: True |1129.204 | 1708.912 |  1.513
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: False | 3655.915  | 7992.877 | 2.186
M: 8192, N: 768, K: 3072, trans_a: False, trans_b: True | 3707.993 |  8026.191 | 2.165
Batch: 768, M: 128, N: 64, K: 128  | 1296.419 | 1308.411 | 1.009

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe commented:
@pytorchbot merge

jgong5 commented:
@pytorchbot merge
